### PR TITLE
DDCNL-7206 - TAI backend Scala 2.13 library upgrades

### PR DESCRIPTION
**Pull request by askar-makkunnath**
- [ ] Unit tests passing
- [ ] Integration tests passing
- [ ] Disabled "-Werror" and the warnings will be fixed in [DDCNL-7682]

### CODE REVIEW

File: it/uk/gov/hmrc/tai/integration/GetEmployeeExpensesSpec.scala [hunk 1]
```diff
@@ -27,7 +27,8 @@ import scala.concurrent.ExecutionContext
 class GetEmployeeExpensesSpec extends IntegrationSpec {
 
   val apiUrl = s"/tai/$nino/tax-account/$year/expenses/employee-expenses/59"
-  def request = FakeRequest(GET, apiUrl).withHeaders("X-SESSION-ID" -> generateSessionId)
+  def request = FakeRequest(GET, apiUrl)
+    .withHeaders("X-SESSION-ID" -> generateSessionId, "AUTHORIZATION" -> "Bearer 11")
```
Comment (pascal22p) [inline it/uk/gov/hmrc/tai/integration/GetEmployeeExpensesSpec.scala:31]: Use HeaderNames.xSessionId and HeaderNames.authorisation

File: it/uk/gov/hmrc/tai/integration/GetEmploymentsSpec.scala [hunk 2]
```diff
@@ -31,7 +31,8 @@ class GetEmploymentsSpec extends IntegrationSpec {
   }
 
   val apiUrl = s"/tai/$nino/employments/years/$year"
-  def request = FakeRequest(GET, apiUrl).withHeaders("X-SESSION-ID" -> generateSessionId)
+  def request = FakeRequest(GET, apiUrl)
+    .withHeaders("X-SESSION-ID" -> generateSessionId, "AUTHORIZATION" -> "Bearer 11")
```
Comment (pascal22p) [inline it/uk/gov/hmrc/tai/integration/GetEmploymentsSpec.scala:35]: Use HeaderNames.xSessionId and HeaderNames.authorisation

File: it/uk/gov/hmrc/tai/integration/PostEmployeeExpensesSpec.scala [hunk 3]
```diff
@@ -30,7 +30,9 @@ class PostEmployeeExpensesSpec extends IntegrationSpec {
 
   val postRequest = Json.toJson(IabdUpdateExpensesRequest(etag.toInt, 123456))
 
-  def request = FakeRequest(POST, apiUrl).withJsonBody(postRequest)
+  def request = FakeRequest(POST, apiUrl)
+    .withJsonBody(postRequest)
+    .withHeaders("AUTHORIZATION" -> "Bearer 11")
```
Comment (pascal22p) [inline it/uk/gov/hmrc/tai/integration/PostEmployeeExpensesSpec.scala:35]: Use HeaderNames.authorisation

File: it/uk/gov/hmrc/tai/integration/PutWorkingFromHomeEmployeeExpenses.scala [hunk 4]
```diff
@@ -30,7 +30,9 @@ class PutWorkingFromHomeEmployeeExpenses extends IntegrationSpec {
 
   val postRequest = Json.toJson(IabdUpdateExpensesRequest(etag.toInt, 123456))
 
-  def request = FakeRequest(POST, apiUrl).withJsonBody(postRequest)
+  def request = FakeRequest(POST, apiUrl)
+    .withJsonBody(postRequest)
+    .withHeaders( "AUTHORIZATION" -> "Bearer 11")
```
Comment (pascal22p) [inline it/uk/gov/hmrc/tai/integration/PutWorkingFromHomeEmployeeExpenses.scala:35]: Use HeaderNames.authorisation

File: it/uk/gov/hmrc/tai/integration/TaxAccountSummarySpec.scala [hunk 5]
```diff
@@ -35,7 +35,8 @@ class TaxAccountSummarySpec extends IntegrationSpec {
   }
 
   val apiUrl = s"/tai/$nino/tax-account/$year/summary"
-  def request = FakeRequest(GET, apiUrl).withHeaders("X-SESSION-ID" -> generateSessionId)
+  def request = FakeRequest(GET, apiUrl)
+    .withHeaders("X-SESSION-ID" -> generateSessionId, "AUTHORIZATION" -> "Bearer 11")
```
Comment (pascal22p) [inline it/uk/gov/hmrc/tai/integration/TaxAccountSummarySpec.scala:39]: Use HeaderNames.xSessionId and HeaderNames.authorisation

File: it/uk/gov/hmrc/tai/integration/TaxCodeChangeSpec.scala [hunk 6]
```diff
@@ -24,7 +24,8 @@ import uk.gov.hmrc.tai.integration.utils.IntegrationSpec
 class TaxCodeChangeSpec extends IntegrationSpec {
 
   val apiUrl = s"/tai/$nino/tax-account/tax-code-change"
-  def request = FakeRequest(GET, apiUrl).withHeaders("X-SESSION-ID" -> generateSessionId)
+  def request = FakeRequest(GET, apiUrl)
+    .withHeaders("X-SESSION-ID" -> generateSessionId, "AUTHORIZATION" -> "Bearer 11")
```
Comment (pascal22p) [inline it/uk/gov/hmrc/tai/integration/TaxCodeChangeSpec.scala:28]: Use HeaderNames.xSessionId and HeaderNames.authorisation

File: it/uk/gov/hmrc/tai/integration/TaxCodeLatestSpec.scala [hunk 7]
```diff
@@ -28,7 +28,8 @@ class TaxCodeLatestSpec extends IntegrationSpec {
 
 
   val apiUrl = s"/tai/$nino/tax-account/$year/tax-code/latest"
-  def request = FakeRequest(GET, apiUrl).withHeaders("X-SESSION-ID" -> generateSessionId)
+  def request = FakeRequest(GET, apiUrl)
+    .withHeaders("X-SESSION-ID" -> generateSessionId, "AUTHORIZATION" -> "Bearer 11")
```
Comment (pascal22p) [inline it/uk/gov/hmrc/tai/integration/TaxCodeLatestSpec.scala:32]: Use HeaderNames.xSessionId and HeaderNames.authorisation

File: it/uk/gov/hmrc/tai/integration/TaxCodeMismatchSpec.scala [hunk 8]
```diff
@@ -33,7 +33,8 @@ class TaxCodeMismatchSpec extends IntegrationSpec {
   }
 
   val apiUrl = s"/tai/$nino/tax-account/tax-code-mismatch"
-  def request = FakeRequest(GET, apiUrl).withHeaders("X-SESSION-ID" -> generateSessionId)
+  def request = FakeRequest(GET, apiUrl)
+    .withHeaders("X-SESSION-ID" -> generateSessionId, "AUTHORIZATION" -> "Bearer 11")
```
Comment (pascal22p) [inline it/uk/gov/hmrc/tai/integration/TaxCodeMismatchSpec.scala:37]: Use HeaderNames.xSessionId and HeaderNames.authorisation

File: it/uk/gov/hmrc/tai/integration/TaxFreeAmountComparisonSpec.scala [hunk 9]
```diff
@@ -24,7 +24,8 @@ import uk.gov.hmrc.tai.integration.utils.IntegrationSpec
 class TaxFreeAmountComparisonSpec extends IntegrationSpec {
 
   val apiUrl = s"/tai/$nino/tax-account/tax-free-amount-comparison"
-  def request = FakeRequest(GET, apiUrl).withHeaders("X-SESSION-ID" -> generateSessionId)
+  def request = FakeRequest(GET, apiUrl)
+    .withHeaders("X-SESSION-ID" -> generateSessionId, "AUTHORIZATION" -> "Bearer 11")
```
Comment (pascal22p) [inline it/uk/gov/hmrc/tai/integration/TaxFreeAmountComparisonSpec.scala:28]: Use HeaderNames.xSessionId and HeaderNames.authorisation

File: it/uk/gov/hmrc/tai/integration/UpdateIncomeSpec.scala [hunk 10]
```diff
@@ -39,7 +39,9 @@ class UpdateIncomeSpec extends IntegrationSpec {
   val amount = 123
   val postRequest = Json.toJson(UpdateTaxCodeIncomeRequest(123456))
 
-  def request = FakeRequest(PUT, apiUrl).withJsonBody(postRequest).withHeaders("X-SESSION-ID" -> generateSessionId)
+  def request = FakeRequest(PUT, apiUrl)
+    .withJsonBody(postRequest)
+    .withHeaders("X-SESSION-ID" -> generateSessionId, "AUTHORIZATION" -> "Bearer 11")
```
Comment (pascal22p) [inline it/uk/gov/hmrc/tai/integration/UpdateIncomeSpec.scala:44]: Use HeaderNames.xSessionId and HeaderNames.authorisation

File: project/AppDependencies.scala [hunk 11]
```diff
@@ -1,39 +1,54 @@
...
-  import play.sbt.PlayImport._
 
   private val playVersion = "play-28"
 
-  private val pegdownVersion = "1.6.0"
-
   private val hmrcMongoVersion = "0.74.0"
 
-  val compile = Seq(
+  private val bootstrapVersion = "7.0.0"
```
Comment (pascal22p) [inline project/AppDependencies.scala:10]: 7.19.0 now?

File: project/AppDependencies.scala [hunk 12]
```diff
@@ -1,39 +1,54 @@
...
 import sbt._
 
 object AppDependencies {
-  import play.sbt.PlayImport._
 
   private val playVersion = "play-28"
 
-  private val pegdownVersion = "1.6.0"
-
   private val hmrcMongoVersion = "0.74.0"
```
Comment (pascal22p) [inline project/AppDependencies.scala:8]: raise a ticket to update this to 1.x

File: project/AppDependencies.scala (hunk 11 above)
Comment (askar-makkunnath) [reply project/AppDependencies.scala]: Updated.

File: it/uk/gov/hmrc/tai/integration/UpdateIncomeSpec.scala (hunk 10 above)
Comment (askar-makkunnath) [reply it/uk/gov/hmrc/tai/integration/UpdateIncomeSpec.scala]: Updated

File: it/uk/gov/hmrc/tai/integration/TaxFreeAmountComparisonSpec.scala (hunk 9 above)
Comment (askar-makkunnath) [reply it/uk/gov/hmrc/tai/integration/TaxFreeAmountComparisonSpec.scala]: Updated

File: it/uk/gov/hmrc/tai/integration/TaxCodeMismatchSpec.scala (hunk 8 above)
Comment (askar-makkunnath) [reply it/uk/gov/hmrc/tai/integration/TaxCodeMismatchSpec.scala]: Updated

File: it/uk/gov/hmrc/tai/integration/GetEmployeeExpensesSpec.scala (hunk 1 above)
Comment (askar-makkunnath) [reply it/uk/gov/hmrc/tai/integration/GetEmployeeExpensesSpec.scala]: Updated

File: it/uk/gov/hmrc/tai/integration/GetEmploymentsSpec.scala (hunk 2 above)
Comment (askar-makkunnath) [reply it/uk/gov/hmrc/tai/integration/GetEmploymentsSpec.scala]: Updated

File: it/uk/gov/hmrc/tai/integration/PostEmployeeExpensesSpec.scala (hunk 3 above)
Comment (askar-makkunnath) [reply it/uk/gov/hmrc/tai/integration/PostEmployeeExpensesSpec.scala]: Updated

File: it/uk/gov/hmrc/tai/integration/PutWorkingFromHomeEmployeeExpenses.scala (hunk 4 above)
Comment (askar-makkunnath) [reply it/uk/gov/hmrc/tai/integration/PutWorkingFromHomeEmployeeExpenses.scala]: Updated

File: it/uk/gov/hmrc/tai/integration/TaxAccountSummarySpec.scala (hunk 5 above)
Comment (askar-makkunnath) [reply it/uk/gov/hmrc/tai/integration/TaxAccountSummarySpec.scala]: Updated

File: it/uk/gov/hmrc/tai/integration/TaxCodeLatestSpec.scala (hunk 7 above)
Comment (askar-makkunnath) [reply it/uk/gov/hmrc/tai/integration/TaxCodeLatestSpec.scala]: Updated

File: app/uk/gov/hmrc/tai/service/TaxFreeAmountComparisonService.scala [hunk 13]
```diff
@@ -71,9 +72,10 @@ class TaxFreeAmountComparisonService @Inject()(
     implicit hc: HeaderCarrier): Future[Seq[CodingComponent]] = {
     val previousCodingComponentsFuture = codingComponentService.codingComponentsForTaxCodeId(nino, taxCodeId)
 
-    previousCodingComponentsFuture.onFailure {
+    previousCodingComponentsFuture.failed.foreach {
       case NonFatal(e) =>
         Logger.error("Could not fetch previous coding components for TaxFreeAmountComparison - " + e.getMessage)
+      case _ => None
```
[the remark this answers is not in the document]
Comment (askar-makkunnath) [reply app/uk/gov/hmrc/tai/service/TaxFreeAmountComparisonService.scala]: Updated by re-throwing exception in 'default' case.

File: app/uk/gov/hmrc/tai/service/TaxFreeAmountComparisonService.scala [hunk 14]
```diff
@@ -50,9 +50,10 @@ class TaxFreeAmountComparisonService @Inject()(
   private def getCurrentComponents(nino: Nino)(implicit hc: HeaderCarrier): Future[Seq[CodingComponent]] = {
     val currentCodingComponentFuture = codingComponentService.codingComponents(nino, TaxYear())
 
-    currentCodingComponentFuture.onFailure {
+    currentCodingComponentFuture.failed.foreach {
       case NonFatal(e) =>
         Logger.error("Could not fetch current coding components for TaxFreeAmountComparison - " + e.getMessage)
+      case _ => None
```
[the remark this answers is not in the document]
Comment (askar-makkunnath) [reply app/uk/gov/hmrc/tai/service/TaxFreeAmountComparisonService.scala]: Updated by re-throwing exception in 'default' case.

File: app/uk/gov/hmrc/tai/service/TaxCodeChangeService.scala [hunk 15]
```diff
@@ -125,9 +124,10 @@ class TaxCodeChangeServiceImpl @Inject()(
       taxCodeMismatch
     }
 
-    futureMismatch.onFailure {
+    futureMismatch.failed.foreach {
       case NonFatal(exception) =>
         logger.warn(s"Failed to compare tax codes for $nino with exception:${exception.getMessage}", exception)
+      case _ => None
```
[the remark this answers is not in the document]
Comment (askar-makkunnath) [reply app/uk/gov/hmrc/tai/service/TaxCodeChangeService.scala]: Updated by re-throwing exception in 'default' case.

File: app/uk/gov/hmrc/tai/model/domain/formatters/TaxAccountSummaryHodFormatters.scala [hunk 16]
```diff
@@ -72,6 +72,7 @@ trait TaxOnOtherIncomeFormatters extends BaseTaxAccountHodFormatters {
           case xs if nonCodedIncome <= xs.head.income =>
             val newTotal = nonCodedIncome * (xs.head.rate / 100)
             total + newTotal
+          case _ => BigDecimal(0)
```
[the remark this answers is not in the document]
Comment (askar-makkunnath) [reply app/uk/gov/hmrc/tai/model/domain/formatters/TaxAccountSummaryHodFormatters.scala]: Updated by re-throwing exception in 'default' case.

File: app/uk/gov/hmrc/tai/calculators/EstimatedPayCalculator.scala [hunk 17]
```diff
@@ -71,6 +71,7 @@ object EstimatedPayCalculator {
         (if (days > 0 && days <= daysInYear) pay * (BigDecimal(daysInYear) / days) else BigDecimal(0))
           .setScale(0, RoundingMode.DOWN)
       }
+      case _ => BigDecimal(0)
```
[the remark this answers is not in the document]
Comment (askar-makkunnath) [reply app/uk/gov/hmrc/tai/calculators/EstimatedPayCalculator.scala]: Updated by re-throwing exception in 'default' case.

File: app/uk/gov/hmrc/tai/connectors/HttpHandler.scala [hunk 18]
```diff
@@ -44,7 +43,7 @@ class HttpHandler @Inject()(metrics: Metrics, httpClient: HttpClient)(implicit e
           case Status.OK =>
             Try(response) match {
               case Success(data) => data
-              case Failure(e)    => throw new RuntimeException("Unable to parse response")
+              case Failure(_)    => throw new RuntimeException("Unable to parse response")
```
[the remark this answers is not in the document]
Comment (askar-makkunnath) [reply app/uk/gov/hmrc/tai/connectors/HttpHandler.scala]: Do we need the try section? can we directly return:
    case Status.OK => response

File: project/AppDependencies.scala (hunk 12 above)
Comment (askar-makkunnath) [reply project/AppDependencies.scala:8]: Created https://jira.tools.tax.service.gov.uk/browse/DDCNL-7702 to track this.

**Review by pascal22p — APPROVED** [review]

**Pull request merged**